/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

edition = "2023";

package cuttlefish.config;

message Graphics {
  // Whether the guest supports the Gfxstream graphics driver.
  bool gfxstream_supported = 1;

  // Whether the guest supports the Gfxstream GL program binary link status
  // feature.
  bool gfxstream_gl_program_binary_link_status_supported = 2;

  // Whether the guest UI can use the BGRA pixel format for its framebuffers.
  bool bgra_framebuffers_supported = 3;

  // If true, the system will prefer using the drm_virgl display mode when
  // available.
  bool prefer_drm_virgl_when_supported = 4;
}

message Audio {
  // The number of output audio streams the guest device supports.
  int32 output_audio_streams_count = 1;
}

message Input {
  // Indicates whether the guest has native support for a mouse.
  bool mouse_supported = 1;

  // Indicates whether the guest has native support for a gamepad or controller.
  bool gamepad_supported = 2;

  // An optional path to a custom keyboard configuration file.
  // The path is relative to the host artifacts directory.
  string custom_keyboard_config = 3;

  // An optional path to a custom DOM key mapping file for keyboard translation.
  // The path is relative to the host artifacts directory.
  string domkey_mapping_config = 4;
}

message Network {
  // If set, forces the use of mac80211_hwsim for Wi-Fi simulation.
  bool enforce_mac80211_hwsim = 1;
}

message Storage {
  // The size in megabytes for a blank userdata.img if one is not provided.
  int32 blank_data_image_mb = 1;
}

message Virtualization {
  // Whether the guest supports vhost-user-vsock for virtualized socket
  // communication.
  bool vhost_user_vsock = 1;

  // An optional path to a Titan C (ti50) security chip emulator binary.
  string ti50_emulator_path = 2;
}

// Specifies the type of the emulated device
enum DeviceType {
    Phone = 0;
    Wear = 1;
    Auto = 2;
    Foldable = 3;
    Tv = 4;
    Minidroid = 5;
    Go = 6;
    Unknown = 7;
}

// GuestConfigFile represents the configuration for a Cuttlefish guest device,
// typically parsed from an cuttlefish-guest-config.txtpb file. It defines the
// device's hardware capabilities, software features, and virtualization
// settings.
message GuestConfigFile {

  DeviceType device_type = 1;

  Graphics graphics = 2;

  Input input = 3;

  Audio audio = 4;

  Network network = 5;

  Storage storage = 6;

  Virtualization virtualization = 7;
}
